Remove save button from entity forms

MKodde opened the pull request:
This PR is aimed to address all tasks found in: https://www.pivotaltracker.com/story/show/166004131

pablothedude commented on src/Surfnet/ServiceProviderDashboard/Domain/Entity/Entity.php:
```diff
@@ -429,180 +408,180 @@ class Entity
-            $entity->{$setter}($attribute);
-        }
-    }
+//    /**
```
Debug leftover?

pablothedude commented on src/Surfnet/ServiceProviderDashboard/Domain/Entity/Entity.php:
```diff
@@ -40,27 +40,6 @@
  */
 class Entity
 {
-    const BINDING_HTTP_POST = 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST';
```
By moving the constants out of this class the usage of `self::` in this class wouldn't work anymore. Please resolve that.

pablothedude commented on src/Surfnet/ServiceProviderDashboard/Domain/ValueObject/Contact.php:
```diff
@@ -45,6 +46,19 @@ class Contact
      */
     private $phone;
 
+    public static function from(?ContactEntity $contact)
```
Please always return a concrete implementation of a Contact or consider using a null object.

pablothedude reviewed (commented):
@MKodde nice work! I've concluded my code review and I will do a functional review later on.

Some additional findings:
`\Surfnet\ServiceProviderDashboard\Infrastructure\DashboardBundle\Controller\EntityControllerTrait::isSaveAction` could be removed
`\Surfnet\ServiceProviderDashboard\Infrastructure\DashboardBundle\Middleware\DoctrineTransactionMiddleware` and config should be removed
When creating a SAML type the 'default' button is still visible.

MKodde commented on src/Surfnet/ServiceProviderDashboard/Application/Service/EntityMergeService.php:
```diff
@@ -0,0 +1,213 @@
+
+    private function determineProtocol(SaveEntityCommandInterface $command)
+    {
+        switch (get_class($command)) {
```
Good point, addressed this in 6614f0d

MKodde commented on src/Surfnet/ServiceProviderDashboard/Application/Service/EntityMergeService.php:
```diff
@@ -0,0 +1,213 @@
+
+    public function mergeEntityCommand(SaveEntityCommandInterface $command, ?ManageEntity $manageEntity = null): ManageEntity
+    {
+        $metaData = new MetaData(
```
Added the methods to the interace in: df92175

MKodde commented on src/Surfnet/ServiceProviderDashboard/Domain/Entity/Entity/Coin.php:
```diff
@@ -121,4 +112,32 @@ public function getOidcClient()
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
+     * @SuppressWarnings(PHPMD.NPathComplexity)
+     */
+    public function merge(?Coin $coin)
```
Making null objects for the value objects of the ManageEntity did cross my mind. But was hesitant to make them in the end. The added benefit seemed small, and the raised complexity during construction time made me decide not to make them.

Or can you change my mind?

pablothedude replied:
Then I would suggest to at least make the vo argument required and move the check to the caller? That way the methods only responsibility is to merge an object. Currently it seems that for example in Metadata the VO's could also be `null` so calling `merge` on them would then fail. This could happen for example when a logourl is an empty string. So if a field could be made required please do so because that will greatly reduce bugs in the future.

MKodde replied:
Yes, that is a good trade-off for now. Venturing into null objects is much easier that way down the road.

pablothedude commented:
Functional review:
- `groups={"production"}` annotation could be removed
- OIDC entity type save button is still there
- Publishing of not-an-instiutute service fails because an entity with institution_id = null is pushed. Could not be null acoording to Manage validation.
- Publishing an OIDC entity fails because the secret is null https://github.com/SURFnet/sp-dashboard/commit/9474616765f6d5e422e2ba7a31387d348f580c9c
- Client secret reset fails because this comparison with a VO fails https://github.com/SURFnet/sp-dashboard/commit/83e033e2d8a0c6d5441f44e8eb9a42eda78f9b42
- On the 'Entrity Idp access' page there is a 'Copy to production' menu item in the header, should that still be there?
- The same applies to the 'Edit IdP whitelist' page

MKodde commented:
Thanks for the functional review. 

> `groups={"production"}` annotation could be removed

I'm not aware that the validation restrictions between the test and production environments have changed. For example, on test you only need to provide a support contact, but on production, all three contacts are required. 

NOOP

>  OIDC entity type save button is still there

Well spotted, removed it!

> Publishing of not-an-instiutute service fails because an entity with institution_id = null is pushed. Could not be null acoording to Manage validation.

Fixed that, the service entity was missing in some circumstances on the ManageEntity.

> Publishing an OIDC entity fails because the secret is null [9474616](https://github.com/SURFnet/sp-dashboard/commit/9474616765f6d5e422e2ba7a31387d348f580c9c)

Thanks, fixed that introducing a NullSecret VO

> Client secret reset fails because this comparison with a VO fails [83e033e](https://github.com/SURFnet/sp-dashboard/commit/83e033e2d8a0c6d5441f44e8eb9a42eda78f9b42)

Fixed, this required quite some changes! Please carefully review that changeset in your re-review.

> On the 'Entrity Idp access' page there is a 'Copy to production' menu item in the header, should that still be there?

I don't think any business rules regarding this have changed.

> The same applies to the 'Edit IdP whitelist' page

I don't think any business rules regarding this have changed.

Myself, I noted that the publication status was not correctly displayed for production entities. The publication state was not interpreted correctly in the EntityService. This was addressed by adding some conditions to when updating the status. Not the most elegant way of fixing this, but should be sufficient for now don't you think?
